pylint/checkers/variables.py

```diff
@@ -979,267 +992,261 @@ def visit_assignname(self, node: nodes.AssignName) -> None:
     def visit_delname(self, node: nodes.DelName) -> None:
         self.visit_name(node)
 
-    # pylint: disable=too-many-branches
     def visit_name(self, node: nodes.Name) -> None:
-        """Check that a name is defined in the current scope"""
         stmt = node.statement()
         if stmt.fromlineno is None:
             # name node from an astroid built from live code, skip
             assert not stmt.root().file.endswith(".py")
             return
 
+        self._undefined_and_used_before_checker(node, stmt)
+        self._loopvar_name(node)
+
+    @utils.check_messages(
+        "undefined-variable", "used-before-assignment", "cell-var-from-loop"
+    )
```

Comment:
```suggestion
```

The decorator should be around for `def visit_name`, not here.

Reply:
Shouldn't it be for both? I seem to remember the decorator being used twice for the same message in inner methods.

Reply:
It's black magic until we read what the function is doing exactly 😄 
```
def check_messages(*messages: str) -> Callable:
    """decorator to store messages that are handled by a checker method"""

    def store_messages(func):
        func.checks_msgs = messages
        return func

    return store_messages
```    
I was thinking it was a function for optimization purpose (to not launch check if the message is not activated). To be fair I don't know what is the purpose now.

Reply:
It is used here to determine if a `visit_xxx` method needs to be called. Thus the decorator only needs to be placed before `visit` methods. https://github.com/PyCQA/pylint/blob/e33bc6377bbbf6135888cdadecdfc3baa8497a24/pylint/utils/ast_walker.py#L19-L22